### PR TITLE
Add ability to run the same content against multiple services (like direct API and MS Graph)

### PR DESCRIPTION
**Pull request by rgregg**

### CODE REVIEW

File: ApiDocs.Console/Auth/BasicAccount.cs [hunk 1]
```diff
@@ -25,19 +25,24 @@
...
 namespace ApiDocs.ConsoleApp.Auth
 {
-    using System;
     using System.Threading.Tasks;
+    using System.Collections.Generic;
     using ApiDocs.Validation;
 
     public class BasicAccount : IServiceAccount
     {
+        public BasicAccount()
```
Comment (daspek) [inline ApiDocs.Console/Auth/BasicAccount.cs]: why did you need to add the default ctor?

File: ApiDocs.Console/Auth/OAuthAccount.cs [hunk 2]
```diff
@@ -29,6 +29,7 @@ namespace ApiDocs.ConsoleApp.Auth
     using ApiDocs.Validation;
     using Newtonsoft.Json;
     using System.Threading.Tasks;
+    using System.Collections.Generic;
```
Comment (daspek) [inline ApiDocs.Console/Auth/OAuthAccount.cs]: sort :)

File: ApiDocs.Console/Program.cs [hunk 3]
```diff
@@ -937,16 +956,48 @@ internal static void WriteValidationError(string indent, ValidationError error)
...
+            Dictionary<string, List<string>> undocumentedProperties = new Dictionary<string, List<string>>();
+            foreach (var props in DiscoveredUndocumentedProperties)
+            {
+                List<string> foundPropertiesOnResource;
+                if (!undocumentedProperties.TryGetValue(props.ResourceName, out foundPropertiesOnResource))
+                {
+                    foundPropertiesOnResource = new List<string>();
+                    undocumentedProperties.Add(props.ResourceName, foundPropertiesOnResource);
+                }
+                if (!foundPropertiesOnResource.Contains(props.PropertyName))
```
Comment (daspek) [inline ApiDocs.Console/Program.cs]: make it a HashSet<string> instead of a List<string> and you don't need to worry about the `contains` check

File: ApiDocs.Validation/Json/JsonRewriter.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    using Newtonsoft.Json.Linq;
+    using System;
+    using System.Linq;
+    using System.Collections.Generic;
+
+    public class JsonRewriter
+    {
+
+        /// <summary>
+        /// Rewrite an input JSON string to adhear to the approriate output schema. This method will use the
```
Comment (daspek) [inline ApiDocs.Validation/Json/JsonRewriter.cs]: typo: adhere

File: ApiDocs.Validation/Json/JsonRewriter.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        private static void ApplyMapToPropertyNames(JContainer source, Dictionary<string, string> rewriteMap)
+        {
+            if (null == source) return;
+
+            foreach (var key in source.ToArray())
+            {
+                JProperty prop = key as JProperty;
+                if (prop != null)
+                {
+                    //Console.WriteLine($"key: {prop.Name}, value-type: {prop.Value.GetType().Name}");
```
Comment (daspek) [inline ApiDocs.Validation/Json/JsonRewriter.cs]: yank?

File: ApiDocs.Validation/MultipartMime/MimeContentType.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        public string MimeType { get; set; }
+
+        public Dictionary<string, string> Arguments { get; private set; }
+
+        public string GetArgumentValue(string key)
+        {
+            string value;
+            if (Arguments.TryGetValueForKey(key, StringComparison.OrdinalIgnoreCase, out value))
```
Comment (daspek) [inline ApiDocs.Validation/MultipartMime/MimeContentType.cs]: set the equalitycomparer on the dictionary when you instantiate it, instead of on the lookup

File: ApiDocs.Validation/MultipartMime/MimeContentType.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                return value;
+            return null;
+        }
+
+        public void SetArgumentValue(string key, string value)
+        {
+            Arguments.SetValueForKey(key, StringComparison.OrdinalIgnoreCase, value);
```
Comment (daspek) [inline ApiDocs.Validation/MultipartMime/MimeContentType.cs]: eq comparer on dictionary

File: ApiDocs.Validation/MultipartMime/MultipartMimeContent.cs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        private static List<MessagePart> ParseBodyIntoParts(string body, string boundary)
+        {
+            var parts = new List<MessagePart>();
+
+            StringReader reader = new StringReader(body);
+            string nextLine = reader.ReadLine();
+            StringBuilder partbody = new StringBuilder();
+            while(nextLine != null)
+            {
+                if (nextLine.Equals($"--{boundary}"))
```
Comment (daspek) [inline ApiDocs.Validation/MultipartMime/MultipartMimeContent.cs]: i'm so sad you had to write a mime parser. there wasn't one available you could use?

File: ApiDocs.Validation/Params/RequestDefinitionExtensions.cs [hunk 9]
```diff
@@ -137,6 +141,73 @@ public static void RewriteRequestWithParameters(this HttpRequest request, IEnume
...
+
+        /// <summary>
+        /// Swap the keys and values in a dictionary (values become keys, keys become values).
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="dictionary"></param>
+        /// <returns></returns>
+        private static Dictionary<T, T> SwapKeysAndValues<T>(this Dictionary<T, T> dictionary)
+        {
+            var output = new Dictionary<T, T>();
```
Comment (daspek) [inline ApiDocs.Validation/Params/RequestDefinitionExtensions.cs]: what a wasted perfect opportunity for an unreadable linq one-liner!
`return dictionary.ToDictionary(kvp=>kvp.Value,kvp=>kvp.Key);`

File: ApiDocs.Console/Auth/BasicAccount.cs (hunk 1 above)
Comment (rgregg) [inline ApiDocs.Console/Auth/BasicAccount.cs]: Not sure... I think that actually came from an older change I made.

File: ApiDocs.Validation/MultipartMime/MimeContentType.cs (hunk 6 above)
Comment (rgregg) [inline ApiDocs.Validation/MultipartMime/MimeContentType.cs]: Oh damn, you can do that?! 👍

File: ApiDocs.Validation/MultipartMime/MultipartMimeContent.cs (hunk 8 above)
Comment (rgregg) [inline ApiDocs.Validation/MultipartMime/MultipartMimeContent.cs]: There probably was, if I we used something other than multipart/related. However finding one that supported that took longer than writing one quickly myself.

File: ApiDocs.Validation/Params/RequestDefinitionExtensions.cs (hunk 9 above)
Comment (rgregg) [inline ApiDocs.Validation/Params/RequestDefinitionExtensions.cs]: That ToDictionary syntax is too hard for me to comprehend in IntelliSense :). Also this whole method is unused now, so I needed to remove this anyway.

File: ApiDocs.Validation/MultipartMime/MimeContentType.cs (hunk 7 above)
Comment (rgregg) [inline ApiDocs.Validation/MultipartMime/MimeContentType.cs]: Cleaned this up in several places, I should have known Dictionary would support that natively.
